File: services/catalog/internal/handlers/health.go
package handlers

import (
	"encoding/json"
	"net/http"
)

// Healthz is a liveness probe. It never checks dependencies — it only
// confirms the process is up and able to handle HTTP requests.
func (h *Handler) Healthz(w http.ResponseWriter, r *http.Request) {
	writeJSON(w, http.StatusOK, map[string]string{"status": "ok"})
}

// Readyz is a readiness probe. It actually pings MongoDB and reports
// 503 with a reason if the dependency is unreachable.
func (h *Handler) Readyz(w http.ResponseWriter, r *http.Request) {
	if err := h.DB.Ping(r.Context()); err != nil {
		h.Logger.Warn("readiness check failed", errField(err))
		writeJSON(w, http.StatusServiceUnavailable, map[string]string{
			"status": "not-ready",
			"reason": err.Error(),
		})
		return
	}
	writeJSON(w, http.StatusOK, map[string]string{"status": "ready"})
}

func writeJSON(w http.ResponseWriter, status int, body any) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	_ = json.NewEncoder(w).Encode(body)
}

File: services/cart/src/routes/health.ts
import type { FastifyInstance } from 'fastify';
import { redisClient } from '../lib/redisClient';

export async function healthRoutes(app: FastifyInstance): Promise<void> {
  // Liveness: no dependency checks, just confirms the process is up.
  app.get('/healthz', async (_req, reply) => {
    return reply.status(200).send({ status: 'ok' });
  });

  // Readiness: must actually check Redis connectivity.
  app.get('/readyz', async (_req, reply) => {
    const redisOk = await redisClient.ping();

    if (!redisOk) {
      return reply.status(503).send({ status: 'not-ready', reason: 'redis unreachable' });
    }

    return reply.status(200).send({ status: 'ready' });
  });
}
